java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/azurecognitivesearch/AzureCognitiveSearchMemory.java

```diff
@@ -0,0 +1,362 @@
+
+        SearchAsyncClient client = this.getSearchClient(normalizeIndexName(collection));
+
+        SearchOptions options =
+                new SearchOptions()
+                        .setQueryType(QueryType.SEMANTIC)
+                        .setSemanticConfigurationName("default")
+                        .setQueryLanguage(QueryLanguage.EN_US)
+                        .setAnswersCount(limit);
+
+        return client.search(query, options).byPage().collect(toMemoryQueryResultList);
+    }
+
+    @Override
+    public Mono<List<String>> getCollectionsAsync() {
+        return this._adminClient
+                .listIndexes()
+                .map(SearchIndex::getName)
+                .collect(Collectors.toList());
+    }
+
+    // Index names cannot contain special chars. We use this rule to replace a few common ones
+    // with an underscore and reduce the chance of errors. If other special chars are used, we leave
+    // it
+    // to the service to throw an error.
+    // Note:
+    // - replacing chars introduces a small chance of conflicts, e.g. "the-user" and "the_user".
+    // - we should consider whether making this optional and leave it to the developer to handle.
+    //
+    // Java 8 does not support \s for whitespace, so we have to list them all.
+    private static String whitespace = " |\\t|\\n|\\x0B|\\f|\\r";
+    private static String s_replaceIndexNameSymbolsRegex = "[" + whitespace + "|\\|/|.|_|:]";
+
+    // Get a search client for the index specified.
+    // Note: the index might not exist, but we avoid checking everytime and the extra latency.
+    // <param name="indexName">Index name</param>
+    // <returns>Search client ready to read/write</returns>
+    private SearchAsyncClient getSearchClient(String indexName) {
+        // Search an available client from the local cache
+        String normalizeIndexName = normalizeIndexName(indexName);
+        return _clientsByIndex.computeIfAbsent(
+                indexName, k -> _adminClient.getSearchAsyncClient(normalizeIndexName));
+    }
+
+    // Create a new search index.
+    // <param name="indexName">Index name</param>
+    // <param name="cancellationToken">Task cancellation token</param>
+    //    private Mono<SearchIndex> createIndexAsync(String indexName) {
+    private Mono<SearchIndex> createIndexAsync(String indexName) {
+
+        indexName = normalizeIndexName(indexName);
+
+        // TODO: FieldBuilder.build(AzureCognitiveSearchRecord.class, null) gives an
+        //  NPE: Cannot invoke "c.f.j.d.i.TypeResoultionContext.resolveType(j.l.r.Type)" because
+        // "this._typeContext" is null
+        //  so we have to do it manually for now.
+        //        List<SearchField> fields = FieldBuilder.build(AzureCognitiveSearchRecord.class,
+        // null);
+        List<SearchField> fields = AzureCognitiveSearchRecord.searchFields();
+
+        SearchIndex newIndex =
+                new SearchIndex(indexName, fields)
```

Comment:
probably should split this over a few lines

Reply:
Ugh. Code formatting at its worst.